File: 2021/Day 9/TS/part1.ts
import fs from 'fs'
import path from 'path'

const file: string = fs.readFileSync(path.resolve(__dirname, '../data.txt'), 'utf-8')
const lines = file.split('\n')

// Calculate
const riskLevels: number[] = []
for (let y = 0; y < lines.length; y++) {
    const line = lines[y]
    for (let x = 0; x < line.length; x++) {
        const n = line[x]
        const surr = [lines?.[y - 1]?.[x], lines?.[y + 1]?.[x], line?.[x - 1], line?.[x + 1]]
        if (!surr.find((val) => val && val <= n)) riskLevels.push(parseInt(n) + 1)
    }
}

// Output

const riskLevel = riskLevels.reduce((sum, val) => sum + val, 0)
console.log(`Risk level: ${riskLevel}`)

File: 2021/Day 09/C++/part1.cpp


// Headers
#include <array>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <vector>

// Namespaces
using namespace std;  // Bad practice, I know

// Main
int main() {
    // Get input
    ifstream file("../data.txt");
    vector<string> lines;
    string line;
    while (getline(file, line)) lines.push_back(line);

    // Find risk level
    vector<int> riskLevels;
    for (size_t y = 0; y < lines.size(); y++) {
        const string line = lines[y];
        for (size_t x = 0; x < line.size(); x++) {
            const char n = line[x];
            if ((y == 0 || lines[y - 1][x] > n) && (y == lines.size() - 1 || lines[y + 1][x] > n) &&
                (x == 0 || lines[y][x - 1] > n) && (x == line.size() - 1 || lines[y][x + 1] > n)) {
                riskLevels.push_back(n - 48 + 1);  // 0 is char 48
            }
        }
    }

    // Calculate sum
    int sum = 0;
    for (auto risk : riskLevels) sum += risk;
    cout << "Risk: " << sum << endl;

    return 0;
}
